torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +475,101 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  // Sleep until a future is available, or there are no futures and we are
+  // shutting down.
+  {
+    std::unique_lock<std::mutex> lock(futureTimeoutMutex_);
+    futureTimeoutCV_.wait(lock);
+  }
+
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    {
+      std::lock_guard<std::mutex> lock{futureMutex_};
+      // Estimate amount of time the first future will time out in, and sleep
+      // for that long.
+      if (futureTimeouts_.empty()) {
+        continue;
```

Comment:
Don't busy spin here, just use some equivalent of infinity (e.g. a year is close enough), and wait in the condvar.